source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -155,49 +156,58 @@ static int heap_alloc_internal_check(int16_t alloc_size)
 static int8_t ns_sector_validate(int *sector_start, int direction)
 {
     int8_t ret_val = -1;
-    int size_start, size_end;
-    if (direction) {
-        //Up
-        size_start = *sector_start++;
-        if (size_start < 0) {
-            sector_start += -(size_start);
-        } else {
-            sector_start += size_start;
-        }
-        size_end = *sector_start;
-    } else {
-        size_end = *sector_start--;
-        if (size_end < 0) {
-            sector_start -= -(size_end);
-        } else {
-            sector_start -= size_end;
-        }
-        size_start = *sector_start;
-    }
-    if (size_start == size_end) {
+    int size_start;
+//    int size_end;
+    int multiplier = direction?1:-1;
+    size_start = *sector_start;
+    sector_start += multiplier;
+    multiplier = size_start<0?-multiplier:multiplier;
```

Comment:
Overly complicated, and badly formatted. Flipping the "direction" variable based on the free status is weird.

You're just wanting to get the size of the block (independent of free status), which is abs(size_start).
I suggest (assuming direction modified to be +1 or -1):

    size = abs(*sector_start);
    sector_start += (1 + size + 1) * direction;

Which I think is fairly self-documenting. (Could someone maybe add a comment somewhere actually explaining the heap format?)

That does assume multiply-by-variable is fast though. I guess it should be on anything except poxy M0 devices. If we cared about slow-multiply devices, then it could be:

    size = abs(*sector_start);
    if (direction > 0) {
            sector_start += 1 + size + 1;
    } else {
            sector_start -= 1 + size + 1;
    }